### PR TITLE
Make variable be immutable in code block node

### PR DESCRIPTION
**Pull request by ke-yu**
### Purpose

This PR makes variables in code block node be immutable. That is, in the same code block node, it is not allowed to change the value of a already-defined variable. E.g.,
```
x = 1;
x = 2;  // variable re-definition error
```
List currently is excluded -- it is allowed to mutate items of a list, but the variable itself is still immutable. For example, following mutation is OK:
```
x = {1, 2, 3};
x[0] = 4;
x["foo"] = "bar";
```
But this is not OK:
```
x = {1, 2, 3};
x = {4, 5, 6};
```

Two kinds of warnings are captured.
  * Recursive reference

  * Variable re-definition

Note the statement that redefines a variable or contains a recursive reference won't be executed. It is shown in the following picture:

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers
@aparajit-pratap 

### FYIs
@Racel @kronz @pboyer

### CODE REVIEW

File: src/Engine/ProtoCore/Utils/CompilerUtils.cs [hunk 1]
```diff
@@ -283,12 +284,20 @@ public static bool PreCompileCodeBlock(Core core, ref ParseParam parseParams, ID
...
+            var warnings = Check(astNodes);
+            parseParams.AppendWarnings(warnings);
+
             if (core.BuildStatus.ErrorCount > 0)
             {
                 parseParams.AppendErrors(core.BuildStatus.Errors);
                 parseParams.AppendWarnings(core.BuildStatus.Warnings);
+            }
+
+            if (parseParams.Errors.Count() > 0 || parseParams.Warnings.Count() > 0)
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/Utils/CompilerUtils.cs]: earlier we were returning `false` only if `errors > 0`

**Comment by aparajit-pratap** [comment]
@ke-yu my only concern is earlier we were returning `false` from `PrecompileCodeBlock` only if there were errors but now you're returning `false` for even warnings? What if the warnings are not variable redefinition warnings?

File: src/Engine/ProtoCore/Utils/CompilerUtils.cs (hunk 1 above)
Comment (ke-yu) [reply src/Engine/ProtoCore/Utils/CompilerUtils.cs]: We should terminate earlier for variable redefinition and recursive reference. I'll update it.

**Comment by ke-yu** [comment]
@aparajit-pratap variable redefinition and recursive reference should be error. I've updated the code and return `false` if there are any errors.

**Comment by aparajit-pratap** [comment]
@ke-yu, we discussed this and thought it would be good to honor the first assignment and just throw a warning for any redefinition. The following is the new behavior we talked about with @kronz and @pboyer as per [this document](https://docs.google.com/document/d/1lOlg7ZUJzE06nxEdZIFux9Y-CVyLpwz81isMc0e1R3k/edit):
image
Could you therefore change it such that it continues to throw a warning and still executes with the first assigned value?

**Comment by ke-yu** [comment]
@aparajit-pratap ok. back to warning.

**Comment by ke-yu** [comment]
@aparajit-pratap Note in this PR I haven't disable the second assignment yet. TBD in the other PR probably.

**Comment by ke-yu** [comment]
@aparajit-pratap done :-)

File: src/Engine/ProtoCore/Utils/CompilerUtils.cs [hunk 2]
```diff
@@ -280,16 +281,19 @@ public static bool PreCompileCodeBlock(Core core, ref ParseParam parseParams, ID
...
             }
-            parseParams.AppendParsedNodes(astNodes);
+
+            // Catch the syntax errors and errors for unsupported 
+            // language constructs thrown by compile expression
+            parseParams.AppendWarnings(core.BuildStatus.Warnings);
+            var warnings = Check(astNodes);
+            parseParams.AppendWarnings(warnings);
+
+            parseParams.AppendParsedNodes(astNodes.Where(n => !n.skipMe));
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/Utils/CompilerUtils.cs]: is `skipMe` an existing property of AST node?

File: src/Engine/ProtoCore/Utils/CompilerUtils.cs [hunk 3]
```diff
@@ -280,16 +281,19 @@ public static bool PreCompileCodeBlock(Core core, ref ParseParam parseParams, ID
...
-                parseParams.AppendErrors(core.BuildStatus.Errors);
-                parseParams.AppendWarnings(core.BuildStatus.Warnings);
                 return false;
             }
-            parseParams.AppendParsedNodes(astNodes);
+
+            // Catch the syntax errors and errors for unsupported 
+            // language constructs thrown by compile expression
+            parseParams.AppendWarnings(core.BuildStatus.Warnings);
+            var warnings = Check(astNodes);
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/Utils/CompilerUtils.cs]: I suggest a more descriptive name for `Check` method, maybe something like `CheckForImmutableVariables`?

File: src/Engine/ProtoCore/Utils/CompilerUtils.cs (hunk 2 above)
Comment (ke-yu) [reply src/Engine/ProtoCore/Utils/CompilerUtils.cs]: Yes

File: src/Engine/ProtoCore/Utils/CompilerUtils.cs (hunk 3 above)
Comment (ke-yu) [reply src/Engine/ProtoCore/Utils/CompilerUtils.cs]: It actually checks immutable variables and recursive reference now. In the future we may add more checking to it.

**Pull request merged**